src/datetime/mod.rs

```diff
@@ -490,24 +490,131 @@ impl<Tz: TimeZone> DateTime<Tz> {
+    #[cfg_attr(docsrs, doc(cfg(any(feature = "alloc", feature = "std"))))]
+    pub fn try_to_rfc2822(&self) -> Option<String> {
         let mut result = String::with_capacity(32);
-        crate::format::write_rfc2822(&mut result, self.naive_local(), self.offset.fix())
-            .expect("writing rfc2822 datetime to string should never fail");
-        result
+        crate::format::write_rfc2822(&mut result, self.naive_local(), self.offset.fix()).ok()?;
+        Some(result)
     }
 
-    /// Returns an RFC 3339 and ISO 8601 date and time string such as `1996-12-19T16:39:57-08:00`.
+    /// Returns an RFC 3339 date and time string such as `1996-12-19T16:39:57-08:00`.
+    /// This is also valid ISO 8601.
+    ///
+    /// # Panics
+    ///
+    /// RFC 3339 is only defined on years 0 through 9999, and panics on dates outside of this range.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// # #![allow(deprecated)]
+    /// # use chrono::{TimeZone, Utc};
+    /// let dt = Utc.with_ymd_and_hms(2023, 6, 10, 9, 18, 25).unwrap();
+    /// assert_eq!(dt.to_rfc3339(), "2023-06-10T09:18:25+00:00");
+    /// ```
     #[cfg(any(feature = "alloc", feature = "std"))]
     #[cfg_attr(docsrs, doc(cfg(any(feature = "alloc", feature = "std"))))]
     #[must_use]
+    #[deprecated(
+        since = "0.4.27",
+        note = "Can panic on years outside of the range 0..=9999. Use `try_to_rfc3339()` or `to_iso8601` instead."
+    )]
     pub fn to_rfc3339(&self) -> String {
-        // For some reason a string with a capacity less than 32 is ca 20% slower when benchmarking.
+        self.try_to_rfc3339().expect("date outside of defined range for rfc3339")
+    }
+
+    /// Returns an RFC 3339 date and time string such as `1996-12-19T16:39:57-08:00`.
+    /// This is also valid ISO 8601.
+    ///
+    /// # Errors
+    ///
+    /// RFC 3339 is only defined on years 0 through 9999, and returns an error on dates outside of
+    /// this range.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// # use chrono::{TimeZone, Utc};
+    /// let dt = Utc.with_ymd_and_hms(2023, 6, 10, 9, 18, 25).unwrap();
+    /// assert_eq!(dt.try_to_rfc3339(), Some("2023-06-10T09:18:25+00:00".to_owned()));
+    ///
+    /// let dt = Utc.with_ymd_and_hms(10_000, 1, 1, 0, 0, 0).unwrap();
+    /// assert_eq!(dt.try_to_rfc3339(), None);
+    /// ```
+    #[cfg(any(feature = "alloc", feature = "std"))]
+    #[cfg_attr(docsrs, doc(cfg(any(feature = "alloc", feature = "std"))))]
+    #[must_use]
+    pub fn try_to_rfc3339(&self) -> Option<String> {
+        let year = self.year();
+        if !(0..=9999).contains(&year) {
+            return None;
+        }
+        Some(self.to_iso8601())
+    }
+
+    /// Returns an ISO 8601 date and time string such as `1996-12-19T16:39:57-08:00`.
+    ///
+    /// Note that standard supports many different format, we choose the one that is compatible
+    /// with RFC 3339 for most common cases.
+    ///
```

Comment:
nitpick: Remove the blank line as the sentence `This format supports...` is extending the idea of the prior sentence.